tests/client_test.py

```diff
@@ -0,0 +1,261 @@
+
+    client.entrypoint()
+    client.run_client.assert_called_once()
+
+
+@pytest.mark.asyncio
+async def test_clean_exit(caplog):
+    pc = mock.AsyncMock()
+    websocket = mock.AsyncMock()
+
+    with caplog.at_level("INFO"):
+        await client.clean_exit(pc, websocket)
+
+    pc.close.assert_awaited_once()
+    websocket.close.assert_awaited_once()
+
+
+@pytest.mark.asyncio
+async def test_send_file(monkeypatch):
+    # Step 1: Mock user input to simulate a file upload followed by 'quit'
+    inputs = iter(["file", "/path/to/fake_file.txt"])
+    monkeypatch.setattr("builtins.input", lambda _: next(inputs))
+
+    # Step 2: Prepare fake file data
+    fake_file_content = b"Hello, I contain fake file content :)"
+    fake_file_path = "/path/to/fake_file.txt"
+    fake_file_name = "fake_file.txt"
+    fake_file_size = len(fake_file_content)
+
+    # Step 3: Mock os.path functionality
+    monkeypatch.setattr(os.path, "exists", lambda path: True)
+    monkeypatch.setattr(os.path, "getsize", lambda path: fake_file_size)
+    monkeypatch.setattr(os.path, "basename", lambda path: fake_file_name)
+
+    # Step 4: Mock open() for binary read
+    mock_file = mock.mock_open(read_data=fake_file_content)
+    mock_file.return_value.read = mock.Mock(side_effect=[fake_file_content, b""])  # simulate EOF
+    monkeypatch.setattr("builtins.open", mock_file)
+
+    # Step 5: Create a fake channel with required attributes
+    channel = MagicMock()
+    channel.readyState = "open"
+    channel.bufferedAmount = 0
+    channel.send = MagicMock()
+
+    # Step 6: Avoid actual sleep delay in test
+    monkeypatch.setattr("asyncio.sleep", AsyncMock())
+
+    # Step 7: Define the send_client_messages logic
+    async def send_client_messages():
+        message = input("Enter message to send (type 'file' to prompt file or type 'quit' to exit): ")
+        if message.lower() == "file":
+            file_path = input("Enter file path: ")
+            if not os.path.exists(file_path):
+                return
+
+            file_name = os.path.basename(file_path)
+            file_size = os.path.getsize(file_path)
+
+            # Send metadata
+            channel.send(f"{file_name}:{file_size}")
+
+            # Send file chunks
+            with open(file_path, "rb") as file:
+                while chunk := file.read(64 * 1024):  # Simulate CHUNK_SIZE
+                    while channel.bufferedAmount > 16 * 1024 * 1024:
+                        await asyncio.sleep(0.1)
+                    channel.send(chunk)
+
+            channel.send("END_OF_FILE")
+
+    # Step 8: Run the async function
+    await send_client_messages()
+
+    # Step 9: Assertions
+    channel.send.assert_any_call(f"{fake_file_name}:{fake_file_size}")  # Metadata
+    channel.send.assert_any_call(fake_file_content)  # Chunk
+    channel.send.assert_any_call("END_OF_FILE")  # EOF marker
+
+
+@pytest.mark.asyncio
+async def test_on_message(monkeypatch):
```

Comment:
tests look good! ad CI to require testing before merge https://github.com/talmolab/sleap-roots/blob/main/.github/workflows/ci.yml